File: ZuluTimeGPS/ZuluTimeGPS/Services/TimeService.swift
import Foundation
import Combine

class TimeService: ObservableObject {
    @Published private(set) var currentTime = Date()

    private var timer: AnyCancellable?

    private static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm:ss"
        return f
    }()

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    // UTC formatted strings
    var utcTimeFormatted: String {
        let f = Self.timeFormatter
        f.timeZone = TimeZone(identifier: "UTC")
        return f.string(from: currentTime)
    }

    var utcDateFormatted: String {
        let f = Self.dateFormatter
        f.timeZone = TimeZone(identifier: "UTC")
        return f.string(from: currentTime)
    }

    // Local formatted strings
    var localTimeFormatted: String {
        let f = Self.timeFormatter
        f.timeZone = .current
        return f.string(from: currentTime)
    }

    var localDateFormatted: String {
        let f = Self.dateFormatter
        f.timeZone = .current
        return f.string(from: currentTime)
    }

    // Timezone offset string (e.g. "+0500" or "-0300")
    var timezoneOffset: String {
        let seconds = TimeZone.current.secondsFromGMT(for: currentTime)
        let hours = seconds / 3600
        let minutes = abs(seconds / 60 % 60)
        let sign = hours >= 0 ? "+" : ""
        return String(format: "%@%02d%02d", sign, hours, minutes)
    }

    init() {
        timer = Timer.publish(every: 1.0, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] date in
                self?.currentTime = date
            }
    }
}

File: zulutime_gps/lib/services/time_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:intl/intl.dart';

class TimeService with ChangeNotifier {
  late Timer _timer;
  DateTime _currentTime = DateTime.now();
  
  DateTime get currentTime => _currentTime;
  
  // UTC time
  DateTime get utcTime => _currentTime.toUtc();
  
  // Formatted times
  String get utcTimeFormatted => DateFormat('HH:mm:ss').format(utcTime);
  String get utcDateFormatted => DateFormat('yyyy-MM-dd').format(utcTime);
  
  String get localTimeFormatted => DateFormat('HH:mm:ss').format(_currentTime);
  String get localDateFormatted => DateFormat('yyyy-MM-dd').format(_currentTime);
  
  // Timezone offset
  String get timezoneOffset {
    final offset = _currentTime.timeZoneOffset;
    final hours = offset.inHours;
    final minutes = offset.inMinutes.remainder(60).abs();
    final sign = hours >= 0 ? '+' : '';
    return '$sign${hours.toString().padLeft(2, '0')}${minutes.toString().padLeft(2, '0')}';
  }
  
  TimeService() {
    _startTimer();
  }
  
  void _startTimer() {
    _timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      _currentTime = DateTime.now();
      notifyListeners();
    });
  }
  
  @override
  void dispose() {
    _timer.cancel();
    super.dispose();
  }
}
